src/signature/spki.rs

```diff
@@ -0,0 +1,197 @@
+// Copyright 2015 Brian Smith.
+//
+// Permission to use, copy, modify, and/or distribute this software for any
+// purpose with or without fee is hereby granted, provided that the above
+// copyright notice and this permission notice appear in all copies.
+//
+// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHORS DISCLAIM ALL WARRANTIES
+// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR
+// ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+// ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+// OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+use der;
+use signature;
+use untrusted;
+
+/// An error that occurs during certificate validation or name validation.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum VerifyWithSPKIError {
+    /// The encoding of some ASN.1 DER-encoded item is invalid.
+    BadDER,
+
+    /// The signature is invalid for the given public key.
+    InvalidSignatureForPublicKey,
+
+    /// The SignatureAlgorithm does not match the algorithm of the SPKI.
+    /// A mismatch could be because of the algorithm (RSA vs DSA, etc) or the
+    /// parameters (ECDSA_p256 vs ECDSA_384, etc).
+    UnsupportedSignatureAlgorithmForPublicKey,
+}
+
+/// Verify the signature `signature` of message `msg` with the public key
+/// `spki_public_key` using the algorithm `alg`.
+/// `signature::verify` is suitable when the public key bytes you have are the
+/// DER ASN.1 of the key. This method is suitable when the public key is in
+/// SubjectPublicKeyInfo form (https://tools.ietf.org/html/rfc5280#section-4.1),
+/// which is a wrapper containing both the key and an identifier.
+///
+/// A common situation where this encoding is encountered is when using public keys
+/// exported by OpenSSL. If you export an RSA or ECDSA public key from a keypair
+/// with `-pubout` and friends, you will get a DER of an SPKI wrapper containing
+/// a public key. You could extract the bitstring of the key itself and use
+/// `signature:;verify`, but that is often inconvient.
+pub fn verify(signature_alg: &Algorithm,
```

Comment:
If I understand you correctly, the current webpki strips the outermost SEQUENCE in a higher-level function and saves the insides, and only the inside parts are passed to the verification function. However, you're also saying *ring*'s new verification function that takes SPKIs needs to strip the SEQUENCE itself. It would be nice if webpki could pass the entire SEQUENCE to the *ring* verification function so that two different functions aren't needed. Am I understanding things correctly?

If so, the webpki code can be changed to something like this:
```rust
...
let spki_begin reader.mark();
try!(der::nested(... der::SEQUENCE, |spki| Ok(())));
let spki_end = reader.mark();
let spki = try!(reader.get_input_between_marks(spki_begin, spki_end));
...
```

Then `spki` would contain the outer later of the sequence. In fact, this logic should probably be encapsulated in `ring::der` as a function called something like `expect_tag_and_get_tlv()`.

Then webpki could just use the public *ring* function, IIUC. WDYT?